Add UI elements used across markdown topics

jrakibi opened the pull request:
This PR introduces components used in topics related to the first module SCRIPT, all of which are located in the `markdown-ui` folder.

**Changes:**
- [x] Added new components for script-related topics.
- [x] Refactored existing files to improve type safety by specifying types more explicitly across components
- [x] Fix navigation bar for topics not showing on mobile and small screens

jrakibi commented:
@IgboPharaoh this is the PR

Emmanuel-Develops commented on components/bitcoin-topics/layouts/TopicLayout.tsx:
```diff
@@ -9,15 +10,24 @@ import { PrevNextLinks } from "../topic/PrevNextLinks"
+import { ChevronRight } from "lucide-react"
 
-const editUrl = (path: string) =>
+const editUrl = (path: string): string =>
```
`editUrl` is never used

jrakibi replied:
nice catch, I added a new button to suggest edits for a topic

Emmanuel-Develops commented on components/bitcoin-topics/layouts/TopicLayout.tsx:
```diff
@@ -9,15 +10,24 @@ import { PrevNextLinks } from "../topic/PrevNextLinks"
-    next?: { path: string; title: string }
-    prev?: { path: string; title: string }
+    next?: { path: string; title: string } | undefined
+    prev?: { path: string; title: string } | undefined
```
same as above

jrakibi commented on app/topics/[...slug]/page.tsx:
```diff
@@ -5,36 +5,28 @@ import { MDXLayoutRenderer } from "pliny/mdx-components"
-        }
-    })
+
+    const imageList =
```
I declared it as JSON to have the flexibility to extend the data model in the future without changing the schema (like if we want to add altText, caption for images, etc.). But currently, we only use the URL

jrakibi commented on components/bitcoin-topics/layouts/TopicLayout.tsx:
```diff
@@ -9,15 +10,24 @@ import { PrevNextLinks } from "../topic/PrevNextLinks"
-    next?: { path: string; title: string }
-    prev?: { path: string; title: string }
+    next?: { path: string; title: string } | undefined
+    prev?: { path: string; title: string } | undefined
```
Fixed

Emmanuel-Develops commented on utils/content-utils.ts:
```diff
@@ -114,8 +114,8 @@ function findPrevTopic(
 export function getTopicData(slug: string): {
-    prev: { path: string; title: string } | undefined
-    next: { path: string; title: string } | undefined
+    prev: { path: string; title: string }
```
```suggestion
    prev?: { path: string; title: string }
    next?: { path: string; title: string }
```
deployment is failing as `prev` and `next` aren't set as optional `?`

jrakibi replied:
ok, added your commit, you can re-deploy now

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/opcode-data-visualizer.tsx:
```diff
@@ -0,0 +1,465 @@
+            </p>
+            <svg
+                ref={svgRef}
+                width="863"
```
on smaller screens, the svg overflows and breaks screen layout, image below
<img width="466" alt="Screenshot 2024-08-31 at 1 31 57 AM" src="https://github.com/user-attachments/assets/9d2c0452-7139-424f-b01c-ac8226e1a5d2">

```suggestion
                width={window.innerWidth < 520 ? "120%" : "100%"}
                height={"auto"}
```
this is a rough draft to get the SVG large enough to be visible, keep aspect ratio, and not break screen layout on small screens. You can try other methods if this doesn't scale as well as you'd prefer.

Also, the `p-4` on the parent container should be removed (maximizing all the space on a small screen 🙃)

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/opcode-explorer.tsx:
```diff
@@ -0,0 +1,449 @@
+    }, [])
+
+    return (
+        <div className="mx-auto py-1 -mx-4 sm:-mx-8 md:-mx-16 lg:-mx-32 xl:-mx-40">
```
```suggestion
        <div className="mx-auto py-1">
```
negative margins breaks screen layout

jrakibi commented:
@Emmanuel-Develops feedback has been addressed in this commit [17372cd](https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/98/commits/17372cdc1b57a88b4c971406ae5d3c09c6673a7a)

IgboPharaoh commented on components/bitcoin-topics/markdown-ui/opcode-explorer.tsx:
```diff
@@ -0,0 +1,449 @@
+    const [selectedOpCode, setSelectedOpCode] = useState<string>(
+        Object.keys(opcodeData)[0]
+    )
+    const [isAsm, setIsAsm] = useState<boolean>(true)
```
A general thing I noticed, when you use a useState and define it's initial value, the useState automatically defines it's type both for the state and the setState so there might be no need for defining its type in angle braces. 

eg: ``const [state, setState] = useState<boolean>(true)`` 

- the state already defines it's type from the initial state passed which is `true` to be a type of boolean
- the setState also defines it's type to be a type of  `React.Dispatch<React.SetStateAction<boolean>>`, notably a setStateAction of the type of <boolean>

Defining a state's type becomes required when you're dealing with complex data structures

IgboPharaoh commented on components/bitcoin-topics/markdown-ui/opcode-explorer.tsx:
```diff
@@ -0,0 +1,449 @@
+    const animationTimeoutRef = useRef<NodeJS.Timeout | null>(null)
+
+    const opCodes = Object.keys(opcodeData)
+    const currentOpCode = opcodeData[selectedOpCode as keyof typeof opcodeData]
```
Might be troublesome to do but I can think you can define the type of selectedOpCode to be a keyof typeof opcodeData, so it only accepts key types of opcodeData and you don't need to coarse the type here.
This can be a future enhancement.

IgboPharaoh commented on components/bitcoin-topics/markdown-ui/script-visualizer/scriptVisualizer.tsx:
```diff
@@ -1,272 +1,336 @@
+    )
 
-    function classNames(...classes: string[]) {
+    function classNames(...classes: string[]): string {
```
I've pointed this out in your previous PRs but possibly you may have skipped as there are alot of reviews but I'll appreciate it if you make this change.
Can you add this `classNames` function to the utils files and use it from there as it is been used in different sections across the codebase.

jrakibi replied:
done

IgboPharaoh commented on components/bitcoin-topics/markdown-ui/stack-simulator.tsx:
```diff
@@ -0,0 +1,214 @@
+}
+
+const opcodes: Opcode[] = [
+    {
```
The object names `OP_ADD`, `OP_EQUAL`,  and `OP_SWAP` have similar function code in the second and third lines of the execute method,
i.e: 
` if (stack.length < 2) return [...stack, "Error: Insufficient items"]`
  `const [b, a] = [stack.pop(), stack.pop()] as [StackItem, StackItem]`
            
 is there a way we can abstract this out and use it across the 3 places 🤔 ?
 Also this should be moved to the data section please.

jrakibi replied:
done

IgboPharaoh commented on components/bitcoin-topics/markdown-ui/stack-simulator.tsx:
```diff
@@ -0,0 +1,214 @@
+                    <input
+                        type="text"
+                        value={input}
+                        onChange={(e) => setInput(e.target.value)}
```
Is it necessary to add the e.preventDefault() as this already work as is 🤔?

jrakibi replied:
The onChange event handler for the input does not trigger a form submission or any default action that needs to be prevented. I think for the current input handling, it is not required

IgboPharaoh commented on components/bitcoin-topics/markdown-ui/transaction-serializer/TransactionDisplay.tsx:
```diff
@@ -90,96 +68,179 @@ const TransactionsDisplay: React.FC<TransactionsDisplayProps> = ({
     }
 
+    const renderItems = (
+        items: any[],
```
Why the use of any here ?

jrakibi replied:
changed to `TransactionItem`

IgboPharaoh reviewed (commented):
Good work on this PR @jrakibi I left a few comments which can quickly be addressed to get this moving

IgboPharaoh commented:
![Screenshot 2024-09-04 at 14 47 02](https://github.com/user-attachments/assets/79742754-3865-44d6-88c0-c57c2a67a775)
![Screenshot 2024-09-04 at 14 47 15](https://github.com/user-attachments/assets/2a18f831-50ae-444a-8a6c-9e9449a5eab6)

- Also, icons tend to become circular and not maintain their original form on light mode, is this intentional by design ?
- On the community page, I think the discord link should have a `target='_blank'` so the user is navigated to a separate screen to join the community and the bdp community page stays active. We don't want them leaving/closing/forgetting that page anytime soon.
- Could you please check out the Bitcoin Stack Simulator in the locking and unlocking section. I tried to add the unlocking_script and locking_script but it's throwing an error of `Error: Non-numeric values`
- Test your knowledge: At the end of the exercise, I think the `Next` button is supposed to navigate you to `P2PK - Pay To Public Key` which is the next section and not back to the `Stack` which is the first subsection under Overview
- P2PK - Pay To Public Key: At the end of the exercise under this section, the user is supposed to be Navigated to the next section which is `P2PKH`
- P2PSH - Pay To Script Hash (BIP 16): Exercise: P2SH address has same issue as P2PK - Pay To Public Key where the Next button doesn't navigate to the next location
- Exercise 2: ASM Converter : - jumps to Project: Build bitcoin Stack and not `P2MS - Pay to Multisig`
- Image on ASM Converter page just after the intro section doesn't load, please investigate
-  P2MS - Pay To Multisig: Exercise under this section has the same issue where the Next button doesn't navigate to the next location which is supposed to be Project: Build bitcoin Stack
- Project: Build bitcoin Stack: Exercise under this section has the same issue where the Next button doesn't navigate to the next location which is supposed to be P2TR but to contribute
- P2TR: is this section supposed to be there since it's empty 🤔?

**Contrast Issues**
- I think we need to change the contrast of the text here as it is not legible, this is for dark mode.
Also, all external link should have a target='_blank' which navigates you to a new page
[component image]
![Screenshot 2024-09-04 at 15 22 14](https://github.com/user-attachments/assets/78232796-3b06-4ce4-ae2d-7c694fb84aa2)
- Opcode pages, similar component has the same issue
- Script Success and failure page
- Standard and Non-Standard page
- P2PSH - Pay To Script Hash (BIP 16) page
- P2MS - Pay To Multisig page ( Standardness VS Validity section ) the author names are not very legible, same component

Contribution guide:
- Contribution guide still have lorem ipsum dolor sit amet..... text fillers, please investigate.

jrakibi commented on components/bitcoin-topics/markdown-ui/opcode-data-visualizer.tsx:
```diff
@@ -0,0 +1,465 @@
+                OP_PUSH data type.
+            </p>
+            <svg
+                ref={svgRef}
```
done, moved it to the public folder

jrakibi commented on components/bitcoin-topics/markdown-ui/opcode-explorer.tsx:
```diff
@@ -0,0 +1,449 @@
+} from "lucide-react"
+import React, { useState, useRef, useEffect, useCallback } from "react"
+
+const opcodeData = {
```
all opcode data is now inside `content/opcode-explorer.tsx`

jrakibi commented:
Thanks @IgboPharaoh for the thorough review.  
I believe all your feedback has been addressed.

- Navigation between topics works fine.  
- The contrast issue is solved (I updated the color theme for all components to maintain a consistent brand color palette).  
- Icons on the `welcome` page are displayed correctly in both light and dark modes.  
- All external links have `target='_blank'` to open in a new tab.  
- Regarding this issue: *"Could you please check out the Bitcoin Stack Simulator in the locking and unlocking section. I tried to add the unlocking_script and locking_script, but it's throwing an error: Non-numeric values"*  I believe you're pushing `OP_2` to the stack instead of the decimal `2`. This is likely causing the issue. I updated the locking and unlocking scripts text to use `2` instead of `OP_2` to avoid confusion.

The only thing left is the contribution guide, but that isn't related to this PR. I'll update its content in the `bitcoin-topics` repo

IgboPharaoh commented:
I appreciate the speedy feedback to the corrections pointed out, everything checks as noted in your comment.
I just noticed two things while doing a final check, please bear with me

- I think we have a typo error here for knowledge
<img width="321" alt="Screenshot 2024-09-05 at 14 32 06" src="https://github.com/user-attachments/assets/eb791682-3251-44f1-a450-28427f65291d">

- Finally, the icons in light mode look quite different to what we have on dark mode, they have circles beneath them. I don't know why this is so but it can be fixed in a future PR, kindly make the text corrections so I speedily approve 🫡.
<img width="897" alt="Screenshot 2024-09-05 at 14 31 37" src="https://github.com/user-attachments/assets/6e282736-4e06-4622-a2b0-a536802f152e">

jrakibi commented:
Nice catch! I’ve fixed it on the bitcoin-topic repo. 

The icons in dark mode are designed to appear like that. we intentionally didn't add the circles in dark mode

IgboPharaoh commented:
LGTM
@Emmanuel-Develops you can merge
